File: api/breakfast/domain.py
"""Pure money + aggregation logic — the server's authority, with no Django deps so
it can be property-tested in isolation. Money is integer piasters everywhere.

This is the Python twin of web/shared's aggregate.ts; both are property-tested to
the same invariant: Σ(per-person totals) == itemsGrandTotal + deliveryFee.
"""
from __future__ import annotations


def egp(pounds: float) -> int:
    return round(pounds * 100)


def money(p: int) -> str:
    sign = '-' if p < 0 else ''
    a = abs(p)
    return f'{sign}{a // 100}.{a % 100:02d}'


def delivery_split(fee: int, n: int) -> list[int]:
    """Split a fixed fee across n submitters so shares sum to `fee` EXACTLY.
    Floor each share, then hand the remainder out one piaster at a time."""
    if n <= 0:
        return []
    base = fee // n
    remainder = fee - base * n  # 0..n-1
    return [base + 1 if i < remainder else base for i in range(n)]


def compute_result(orders: list[dict], menu: list[dict], delivery_fee: int) -> dict:
    """orders: [{person, paid, lines:[{item_id, qty, unit_price, note, for_name}]}]
    menu:   [{id, name, arabic, kind, sort_order}] (already in display order)
    Returns a CloseResult dict with camelCase keys matching the API contract."""
    position = {m['id']: i for i, m in enumerate(menu)}
    by_id = {m['id']: m for m in menu}

    # vendor rollup: qty per item + notes with counts
    agg: dict[str, dict] = {}
    for o in orders:
        for l in o['lines']:
            item = by_id.get(l['item_id'])
            if not item:
                continue
            a = agg.get(l['item_id'])
            if a is None:
                a = {'itemId': l['item_id'], 'name': item['name'], 'arabic': item.get('arabic'), 'qty': 0, 'notes': []}
                agg[l['item_id']] = a
            a['qty'] += l['qty']
            note = l.get('note')
            if note:
                existing = next((x for x in a['notes'] if x['note'] == note), None)
                if existing:
                    existing['count'] += l['qty']
                else:
                    a['notes'].append({'note': note, 'count': l['qty']})
    aggregate = sorted(agg.values(), key=lambda a: position.get(a['itemId'], 0))

    headcount = len(orders)  # submitters = delivery denominator
    shares = delivery_split(delivery_fee, headcount)

    per_person = []
    for i, o in enumerate(orders):
        items_total = sum(l['qty'] * l['unit_price'] for l in o['lines'])
        delivery_share = shares[i] if i < len(shares) else 0
        for_names = list(dict.fromkeys(l['for_name'] for l in o['lines'] if l.get('for_name')))
        per_person.append({
            'person': o['person'],
            'itemsTotal': items_total,
            'deliveryShare': delivery_share,
            'total': items_total + delivery_share,
            'paid': o['paid'],
            'forNames': for_names,
        })

    items_grand_total = sum(p['itemsTotal'] for p in per_person)
    return {
        'aggregate': aggregate,
        'itemsGrandTotal': items_grand_total,
        'deliveryFee': delivery_fee,
        'grandTotal': items_grand_total + delivery_fee,
        'perPerson': per_person,
        'headcount': headcount,
    }


def verify_exact(result: dict) -> bool:
    sum_people = sum(p['total'] for p in result['perPerson'])
    sum_shares = sum(p['deliveryShare'] for p in result['perPerson'])
    return sum_people == result['grandTotal'] and sum_shares == result['deliveryFee']

File: api/src/lib/domain.ts
import {
  deliverySplit,
  rollup,
  type CloseResult,
  type Restaurant as SharedRestaurant,
  type Session as SharedSession,
} from '@morningcart/shared'
import type { RestaurantFull, SessionFull } from '../db'
import { env } from '../env'

/** YYYY-MM-DD in office-local time. */
export function serviceDateToday(): string {
  const fmt = new Intl.DateTimeFormat('en-CA', {
    timeZone: env.OFFICE_TZ,
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
  })
  return fmt.format(new Date())
}

export function toSharedRestaurant(r: RestaurantFull): SharedRestaurant {
  return {
    id: r.id,
    name: r.name,
    arabic: r.arabic ?? undefined,
    deliveryFee: r.deliveryFee,
    menu: r.menu.map((m) => ({
      id: m.id,
      name: m.name,
      arabic: m.arabic ?? undefined,
      price: m.price,
      kind: m.kind as SharedRestaurant['menu'][number]['kind'],
      available: m.available,
    })),
  }
}

export function toSharedSession(s: SessionFull): SharedSession {
  return {
    id: s.id,
    restaurantId: s.restaurantId,
    startedBy: s.startedBy,
    status: s.status as SharedSession['status'],
    dateLabel: s.serviceDate,
    orders: s.orders.map((o) => ({
      id: o.id,
      person: o.person,
      paid: o.paid,
      lines: o.lines.map((l) => ({
        itemId: l.menuItemId,
        qty: l.qty,
        note: l.note ?? undefined,
        forName: l.forName ?? undefined,
      })),
    })),
  }
}

/**
 * Authoritative close computation. Reuses the shared, canonical, price-independent
 * pieces (rollup for the vendor list, deliverySplit for the fee) but computes each
 * person's itemsTotal from the SNAPSHOTTED OrderLine.unitPrice, so a later menu
 * price edit can never alter a historical bill.
 * Invariant: Σ(perPerson.total) === itemsGrandTotal + deliveryFee, to the piaster.
 */
export function computeResult(session: SessionFull, restaurant: RestaurantFull): CloseResult {
  const aggregate = rollup(toSharedSession(session), toSharedRestaurant(restaurant))
  const orders = session.orders
  const headcount = orders.length // submitters = delivery denominator
  const shares = deliverySplit(restaurant.deliveryFee, headcount)

  const perPerson = orders.map((o, i) => {
    const itemsTotal = o.lines.reduce((sum, l) => sum + l.qty * l.unitPrice, 0)
    const deliveryShare = shares[i] ?? 0
    const forNames = [...new Set(o.lines.map((l) => l.forName).filter((x): x is string => !!x))]
    return {
      person: o.person,
      itemsTotal,
      deliveryShare,
      total: itemsTotal + deliveryShare,
      paid: o.paid,
      forNames,
    }
  })

  const itemsGrandTotal = perPerson.reduce((s, p) => s + p.itemsTotal, 0)
  return {
    aggregate,
    itemsGrandTotal,
    deliveryFee: restaurant.deliveryFee,
    grandTotal: itemsGrandTotal + restaurant.deliveryFee,
    perPerson,
    headcount,
  }
}
